src/goto-instrument/contracts/assigns.cpp

```diff
@@ -112,13 +111,30 @@ void assigns_clauset::add_target(const exprt &target_expr)
 
 void assigns_clauset::remove_target(const exprt &target_expr)
 {
-  targets.erase(targett(*this, targett::normalize(target_expr)));
+  write_set.erase(targett(*this, target_expr));
+}
+
+void assigns_clauset::add_freely_assignable_expr(const exprt &expr)
+{
+  auto insertion_succeeded = freely_assignable_set.emplace(*this, expr).second;
+
+  if(!insertion_succeeded)
+  {
+    log.warning() << "Ignored duplicate expression '"
+                  << from_expr(ns, expr.id(), expr) << "' in assigns clause at "
+                  << expr.source_location().as_string() << messaget::eom;
+  }
+}
+
+void assigns_clauset::remove_freely_assignable_expr(const exprt &expr)
+{
+  freely_assignable_set.erase(targett(*this, targett::normalize(expr)));
 }
 
 goto_programt assigns_clauset::generate_havoc_code() const
 {
   modifiest modifies;
-  for(const auto &target : targets)
+  for(const auto &target : write_set)
     modifies.insert(target.address.object());
```

Comment:
For completeness, we should also havoc the `freely_assignable_set`. So we need a similar loop for that.

Reply:
Implemented as suggested.